File: 신혜원/week3/programmers_이중우선순위큐.js
/*
힙으로 구현하는 것이 효율적이지만 간단히 배열로 구현할 수도 있을거 같아서 배열로 구현.
다음번에는 힙으로 구현할 예정.
*/
function solution(operations) {
    let queue = [];

    for (let op of operations) {
        let [command, value] = op.split(' ');
        value = parseInt(value);

        if (command === 'I') {
            queue.push(value);
        } 
        else if (command === 'D') {
            if (queue.length === 0) continue;

            queue.sort((a, b) => a - b); // 정렬

            if (value === 1) {
                queue.pop();     // 최대값
            } else {
                queue.shift();   // 최소값
            }
        }
    }

    if (queue.length === 0) return [0, 0];

    queue.sort((a, b) => a - b);
    return [queue[queue.length - 1], queue[0]];
}

File: 채주혁/4주차/programmers_이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        TreeMap<Integer, Integer> map = new TreeMap<>();

        for (String op : operations) {
            String[] parts = op.split(" ");
            String command = parts[0];
            int num = Integer.parseInt(parts[1]);

            // 삽입
            if (command.equals("I")) {
                map.put(num, map.getOrDefault(num, 0) + 1);
            }
            // 삭제
            else {
                if (map.isEmpty()) continue;

                // 최댓값 삭제
                if (num == 1) {
                    int maxKey = map.lastKey();
                    if (map.get(maxKey) == 1) {
                        map.remove(maxKey);
                    } else {
                        map.put(maxKey, map.get(maxKey) - 1);
                    }
                }
                // 최솟값 삭제
                else if (num == -1) {
                    int minKey = map.firstKey();
                    if (map.get(minKey) == 1) {
                        map.remove(minKey);
                    } else {
                        map.put(minKey, map.get(minKey) - 1);
                    }
                }
            }
        }

        if (map.isEmpty()) {
            return new int[]{0, 0};
        }

        return new int[]{map.lastKey(), map.firstKey()};
    }
}